go/worker/storage/committee/checkpoint_sync.go

```diff
@@ -0,0 +1,268 @@
+		go func(clientConn *committee.ClientConnWithMeta) {
+			defer workerGroup.Done()
+			n.nodeWorker(workCtx, clientConn, chunkDispatchCh, chunkReturnCh, errorCh)
+		}(clientConn)
+	}
+	go func() {
+		workerGroup.Wait()
+		close(workersDone)
+	}()
+
+	if err := n.localStorage.Checkpointer().StartRestore(n.ctx, check); err != nil {
+		return checkpointStatusNext, fmt.Errorf("can't start checkpoint restore: %w", err)
+	}
+
+	// Prepare the heap of chunks.
+	chunks := &chunkHeap{
+		array: make([]*checkpoint.ChunkMetadata, len(check.Chunks)),
+		length: 0,
+	}
+	heap.Init(chunks)
+
+	for i, c := range check.Chunks {
+		heap.Push(chunks, &checkpoint.ChunkMetadata{
+			Version: 1,
+			Index: uint64(i),
+			Digest: c,
+			Root: check.Root,
+		})
+	}
+
+	// Feed the workers with chunks.
+	var next *checkpoint.ChunkMetadata
+	var outChan chan *checkpoint.ChunkMetadata
+
+	for {
+		if chunks.length == 0 {
+			next = nil
+			outChan = nil
+		} else {
+			next = heap.Pop(chunks).(*checkpoint.ChunkMetadata)
+			outChan = chunkDispatchCh
+		}
+
+		select {
+		case <-n.ctx.Done():
+			return checkpointStatusBail, n.ctx.Err()
+
+		case returned := <-chunkReturnCh:
+			if returned == nil {
+				// Restoration completed, no more chunks.
+				return checkpointStatusDone, nil
+			}
+			heap.Push(chunks, returned)
+
+		case status := <-errorCh:
+			return status, nil
+
+		case outChan <- next:
+			next = nil
+
+		case <-workersDone:
+			// No usable committee connections left, bail.
+			return checkpointStatusBail, storageClient.ErrStorageNotAvailable
+		}
+
+		if next != nil {
+			heap.Push(chunks, next)
+		}
+	}
+}
+
+func (n *Node) syncCheckpoint() error {
+	// Fetch metadata first from the current committee.
+	req := &checkpoint.GetCheckpointsRequest{
+		Namespace: n.commonNode.Runtime.ID(),
+	}
+	metadata, err := n.storageClient.GetCheckpoints(n.ctx, req)
+	if err != nil {
+		return fmt.Errorf("can't get checkpoint list: %w", err)
+	}
+	sort.Slice(metadata, func(i, j int) bool {
+		// Descending!
+		return metadata[j].Root.Version < metadata[i].Root.Version
+	})
+	n.logger.Debug("checkpoints to comb through", "count", len(metadata))
+
+	// Start following the storage committee.
+	committeeWatcher, err := committee.NewWatcher(
+		n.ctx,
+		n.commonNode.Consensus.Scheduler(),
+		n.commonNode.Consensus.Registry(),
+		n.commonNode.Runtime.ID(),
+		schedulerApi.KindStorage,
+		committee.WithAutomaticEpochTransitions(),
+	)
+	if err != nil {
+		return fmt.Errorf("can't establish storage committee watcher: %w", err)
+	}
+	committeeClient, err := committee.NewClient(n.ctx, committeeWatcher.Nodes(), committee.WithClientAuthentication(n.commonNode.Identity))
+	if err != nil {
+		return fmt.Errorf("can't create committee client: %w", err)
+	}
+
+	// Try all the checkpoints now, from most recent backwards.
+	for _, check := range metadata {
+		if !n.localStorage.NodeDB().HasRoot(check.Root) {
+			n.logger.Debug("checkpoint doesn't have a known root", "root", check.Root)
```

Comment:
Hm, this sounds wrong, you will only sync checkpoints for which you already have all the data?

You probably wanted to use `n.commonNode.Runtime.History().GetBlock` to check if you have the block and at the same time, you should filter out all the checkpoints which are lower or equal than `n.syncedState.LastBlock.Round` (because for those you already have all the data synced).